### ISSUE TITLE
Performance improvement: `AreaDefinition.__getitem__` force to lons/lats array recomputations after slicing

### ISSUE THREAD

**Issue by ghiggi** [pytroll/pyresample#411]
#### Problem description 

Currently, the method `AreaDefinition.__getitem__` [(L2436 of geometry.py)](https://github.com/pytroll/pyresample/blob/main/pyresample/geometry.py#L2436), when performing AreaDefinition slicing, it discards the possible precomputed lons/lats arrays attached to the class instance. 
In the case of a large AreaDef (i.e. geostationary), recomputation of lons/lats after a slicing can take several seconds if the sliced domain is quite large. 

#### Proposition of code to add to `AreaDefinition.__getitem__`
```python
if self.lats is not None: 
     lats = lats[yslice,  xslice]
else: 
     lats = None 
if self.lons is not None: 
     lons = lons[yslice,  xslice]
else: 
     lons = None 

# Pass lats, lons to the AreaDefinition call 
 new_area = AreaDefinition(self.area_id, self.description,
                                  self.proj_id, self.crs,
                                  total_cols,
                                  total_rows,
                                  new_area_extent,
                                  lons = lons, lats = lats) # NEW 
```
#### Side note 

The same approach could be employed to also avoid the recomputation of the `cartesian_coords` attribute.

**Comment by djhoese** [comment]
Has this been updated in some PR and is not occurring anymore? If so, which PR?

**Comment by ghiggi** [comment]
Not sure which PR, but now [`the__getitem__`]((https://github.com/pytroll/pyresample/blob/main/pyresample/geometry.py#L125))method is inherited from `BaseDefinition`, which slice the cached lons/lats if available.

**Comment by ghiggi** [comment]
Ouch I was wrong ... still occurring. I can do a PR for that next week ! 
See [here](https://github.com/pytroll/pyresample/blob/main/pyresample/geometry.py#L2626)

**Comment by djhoese** [comment]
Looking at the code I'm not sure I understand the issue. Oh looking at the original code in this issue, you're saying if someone did `area.get_lonlats()` then sliced the area, that we could reuse the original cached lon/lats when we create the AreaDefinition. Right?

What's the use case and do you have any idea what the difference in time is?

**Comment by ghiggi** [comment]
Exactly. I don't remember if it was one of my use case, or something related maybe to angles computations? 
Anyway it's a quick fix that I can do that will eventually  slightly improve the performance ;)

**Comment by djhoese** [comment]
It would help and be fast to do it for dask-based operations, but that is not the default for `get_lonlats`. For numpy arrays it would probably provide the most benefit since hopefully it is the same amount of memory. I was just thinking, if it isn't already, the lon/lats when given to an AreaDefinition should be made read-only.

I'll be honest, I have not thought about this use case when designing/thinking about Pyresample 2.0's area definitions. I really wanted to remove as much of the extra `__init__` arguments as possible. I suppose since it is creating a new instance of itself, the `AreaDefinition` could do `new_area._lons = sliced_lons`.

**Comment by ghiggi** [comment]
Do you mean having lats/lons read-only because are modified somewhere afterward? Do you have a case in mind? We could still do a copy (instead of recomputing it ... in case).
Concerning the AreaDefinition interface in pyresample 2.0 I agree to remove as much as extra arguments in the `__init__`. 
If you aim I do the PR, I can assigns the sliced lat/lon after the new Area Definition

**Comment by djhoese** [comment]
I don't have a case of pyresample itself modifying the lon/lats, but users could take the results and modify them. We don't want to make a copy of the array because of the performance issues it would cause. But we could do:

```
array.flags.writable = False
```

To make it so `__setitem__` operations fail if someone doesn't first make a copy. We wouldn't want someone modifying the internal cache of the AreaDefinition or the contained lon/lats of the SwathDefinition. At least in my opinion this is a good idea.

Ignore Pyresample 2.0 stuff in any work that is done on this.

**Comment by ghiggi** [comment]
I weren't aware of `array.flags.writable = False`. Nice to know ;) But this only works for numpy array right?

**Comment by djhoese** [comment]
Correct. We don't have much choice with dask, but the dask arrays themselves are immutable. The results of a dask computation are writable also I think, but I think we'd be caching the dask array in that case so I'm less concerned. I'm also not aware of something in xarray (but I haven't checked either).
